### ISSUE TITLE
Model training for time-dependent sink and source terms

### ISSUE THREAD

**Issue by Rdfing** [lululxvi/deepxde#369]
Hi everyone,

Thank you Lu Lu for the generous sharing of your source code. 

I am solving PDE with time-dependent sink and source terms. This is relatively straightforward for a specific case as the time-dependent sink and source terms can be embedded in the loss function.

However, I am intended to train a general model such that time-dependent sink and source are part of the model input parameters.  Once the model is trained, the model can be used to predict for any given time-dependent sink and source input. I felt this is achievable (perhaps, I should start with ODE) but I am stuck with formulating the problem.  Is this something achievable?

Thanks in advance,
Haochen

**Comment by lululxvi** [comment]
It seems doable, but I need more details to give you more specific suggestions.

**Comment by Rdfing** [comment]
Lu lu,

Please ignore my question. I wrapped my head around. 

I basically work on a problem that is similar to one of your papers "Systems biology informed deep learning for inferring parameters and hidden dynamics". The only difference is that there are time-dependent sinks and sources that represent system input and output. How many points typically would you use to force such an ODE system? 

Thanks,
Haochen

**Comment by Rdfing** [comment]
Lu lu,

I notice in the Lotka Volterra example, the training completes in 383.3 s, which is significantly faster than what I experienced. Most of my computing resources are CPU for CFD simulations, e.g., AMD EPYC 7702 Rome
2.0 GHz Cores.

I also notice that for my problem, it seems that I can only avoid the early termination problem in LBFGS with float64. What is hardware would you recommend for general PINNs training?

Thanks,
Haochen

**Comment by lululxvi** [comment]
Usually GPU training is faster than CPU, but the speed-up depends on the specific problem. You may have to try both CPU and GPU and see what the difference is.

For L-BFGS, float64 is usually required.

**Comment by engsbk** [comment]
> Hi everyone,
> 
> Thank you Lu Lu for the generous sharing of your source code.
> 
> I am solving PDE with time-dependent sink and source terms. This is relatively straightforward for a specific case as the time-dependent sink and source terms can be embedded in the loss function.
> 
> However, I am intended to train a general model such that time-dependent sink and source are part of the model input parameters. Once the model is trained, the model can be used to predict for any given time-dependent sink and source input. I felt this is achievable (perhaps, I should start with ODE) but I am stuck with formulating the problem. Is this something achievable?
> 
> Thanks in advance, Haochen

I'm also trying to train a model where the time-dependent source function is part of the model. The loss values checks out fine around 10e-5 or 10e-6, however the results are not very accurate in comparison to the FDTD (exact) solution. Did you work around that?

**Comment by Rdfing** [comment]
engsbk

> > Hi everyone,
> > Thank you Lu Lu for the generous sharing of your source code.
> > I am solving PDE with time-dependent sink and source terms. This is relatively straightforward for a specific case as the time-dependent sink and source terms can be embedded in the loss function.
> > However, I am intended to train a general model such that time-dependent sink and source are part of the model input parameters. Once the model is trained, the model can be used to predict for any given time-dependent sink and source input. I felt this is achievable (perhaps, I should start with ODE) but I am stuck with formulating the problem. Is this something achievable?
> > Thanks in advance, Haochen
> 
> I'm also trying to train a model where the time-dependent source function is part of the model. The loss values checks out fine around 10e-5 or 10e-6, however the results are not very accurate in comparison to the FDTD (exact) solution. Did you work around that?

I solved my issue by (1) scaling the input and output features and (2) training the model with L-BFGS on float64.

**Comment by engsbk** [comment]
By scaling, do you mean using:
```net.apply_output_transform(lambda x, u: u * np.array([100]))```

**Comment by Rdfing** [comment]
> By scaling, do you mean using something like: `net.apply_output_transform(lambda x, u: u * np.array([100]))` ?

Just conventional ML feature scaling. I scaled my input and output with their standard deviations. You can just use `sklearn.preprocessing.StandardScaler`

**Comment by engsbk** [comment]
Thank you for your prompt responses. Could you explain with an example, please?
The PDE I'm trying to model although simple but still won't reflect reasonable results.
